actors/evm/src/interpreter/instructions/ext.rs

```diff
@@ -48,40 +58,61 @@ pub fn extcodecopy(
     data_offset: U256,
     size: U256,
 ) -> Result<(), StatusCode> {
-    let bytecode =
-        get_evm_bytecode_cid(system.rt, addr).and_then(|cid| get_evm_bytecode(system.rt, &cid))?;
+    let bytecode = match get_cid_type(system.rt, addr) {
+        CodeCid::EVM(addr) => get_evm_bytecode(system.rt, &addr)?,
+        CodeCid::NotFound | CodeCid::Account => Vec::new(),
+        // calling EXTCODECOPY on native actors results with a single byte 0xFE which solidtiy uses for its `assert`/`throw` methods
+        // and in general invalid EVM bytecode
+        _ => vec![0xFE],
+    };
 
     copy_to_memory(&mut state.memory, dest_offset, size, data_offset, bytecode.as_slice(), true)
 }
 
-pub fn get_evm_bytecode_cid(rt: &impl Runtime, addr: U256) -> Result<Cid, StatusCode> {
-    let addr: EthAddress = addr.into();
-    let addr: Address = addr.try_into()?;
-    // TODO: just return none in most of these cases?
-    let actor_id = rt.resolve_address(&addr).ok_or_else(|| {
-        StatusCode::InvalidArgument("failed to resolve address".to_string())
-        // TODO better error code
-    })?;
-
-    let evm_cid = rt.get_code_cid_for_type(Type::EVM);
-    let target_cid = rt.get_actor_code_cid(&actor_id);
+#[derive(Debug)]
+pub enum CodeCid {
+    /// EVM Address and the CID of the actor (not the bytecode)
+    EVM(Address),
+    Native(Cid),
+    Account,
+    NotFound,
+}
 
-    if Some(evm_cid) != target_cid {
-        return Err(StatusCode::InvalidArgument(
-            "cannot invoke EXTCODESIZE for non-EVM actor".to_string(),
-        )); // TODO better error code
+impl CodeCid {
+    pub fn unwrap_evm<E>(&self, err: E) -> Result<Address, E> {
```

Comment:
`as_evm` (`unwrap` panics)

Reply:
Or `try_unwrap_evm` .

Or just remove it given that we call it once.

Reply:
im for removing. was more useful in previous iterations (and naming was off as mentioned)